finality-aleph/src/new_network/split.rs

```diff
@@ -0,0 +1,231 @@
+use crate::new_network::{ComponentNetwork, Data, ReceiverComponent, SendError, SenderComponent};
+use aleph_bft::Recipient;
+use codec::{Decode, Encode};
+use futures::channel::mpsc;
+use log::warn;
+use std::{marker::PhantomData, sync::Arc};
+use tokio::{
+    sync::Mutex,
+    time::{interval, Duration},
+};
+
+const MAXIMUM_RETRY_MS: u64 = 500;
+
+/// Used for routing data through split networks.
+#[derive(Clone, Encode, Decode)]
+pub enum Split<D1: Data, D2: Data> {
+    Left(D1),
+    Right(D2),
+}
+
+#[derive(Clone)]
+struct LeftSender<D1: Data, D2: Data, S: SenderComponent<Split<D1, D2>>> {
```

Comment:
What will be a name when we will need a third sender? :thinking:

Reply:
If I get it right for example `LeftSender<D1,D2,S> = RightSender<D2, D1, S>`, so I don't see why we should have 2 (4) different structures. I see that they have to send different `enum Split`, but maybe you can somehow work around it

Reply:
Oh, I tried, believe me.. I don't think it's possible to remove much of the duplication in Rust unfortunately. :/

As for your original question – we don't need to, because trees! If we need to split into 3 we split once, and split one of the outputs again. :3 Abstraction!

Reply:
it probably could be deduplicated with macro usage but I don't think it is worth it (or more readable) to do it.